File: src/game.c
typedef Vector2 V2;

typedef struct Body {
    f32 mass, radius;
    V2 pos, velocity;
    Color colour;
} Body;
typedef Array(Body) Array_Body;

typedef struct Creator {
    bool active;
    Body body;
    V2 displacement;
} Creator;

typedef struct Context {
    Arena arena;
    char *name;
    int width, height, fps;

    f32 G, delta, cursor_radius;
    Array_Body bodies;
    Creator creator;
    V2 mouse_pos;
} Context;

static inline f32 scale(Context *ctx) { return (f32)ctx->height; }
static f32 normal_from_screen(Context *ctx, f32 n) { return n / scale(ctx); }
static f32 screen_from_normal(Context *ctx, f32 n) { return n * scale(ctx); }
static f32 normal_width(Context *ctx) { return (f32)ctx->width / scale(ctx); }

static void game_init(Context *ctx) {
    SetConfigFlags(FLAG_MSAA_4X_HINT | FLAG_WINDOW_RESIZABLE);
    SetTargetFPS(ctx->fps);
    InitWindow(ctx->width, ctx->height, ctx->name);
    ctx->G = 3E-8f / (f32)ctx->fps;
    ctx->creator.body.colour = colour_body;
}

static void game_deinit(Context *ctx) { discard(ctx); CloseWindow(); }

static void game_frame_begin(Context *ctx) {
    discard(ctx);
    BeginDrawing();
    ClearBackground(colour_background);
}

static void game_frame_end(Context *ctx) { discard(ctx); EndDrawing(); }

static void game_compute_interaction(Context *ctx, usize i, usize cmp_i) {
    Body *body = &ctx->bodies.ptr[i];
    Body *body_cmp = &ctx->bodies.ptr[cmp_i];

    f32 x_dist = body->pos.x - body_cmp->pos.x;
    f32 y_dist = body->pos.y - body_cmp->pos.y;
    f32 dist = sqrtf(x_dist * x_dist + y_dist * y_dist);

    bool colliding = dist < (body->radius + body_cmp->radius) / 2.0f;
    if (colliding) return;

    f32 factor = -1 * ctx->delta * ctx->G;
    f32 force = factor * body->mass * body_cmp->mass / (dist * dist);
    f32 force_x = force * (x_dist / dist);
    f32 force_y = force * (y_dist / dist);

    V2 body_accel = { force_x / body->mass, force_y / body->mass };
    body->velocity.x += body_accel.x;
    body->velocity.y += body_accel.y;

    V2 body_cmp_accel = { force_x / body_cmp->mass, force_y / body_cmp->mass };
    body_cmp->velocity.x -= body_cmp_accel.x;
    body_cmp->velocity.y -= body_cmp_accel.y;
}

static V2 V2_scale(V2 v, f32 n)  { return (V2){ v.x * n, v.y * n }; }
static V2 V2_divide(V2 v, f32 n) { return (V2){ v.x / n, v.y / n }; }
static V2 V2_add(V2 v1, V2 v2)   { return (V2){ v1.x + v2.x, v1.y + v2.y }; }
static V2 V2_sub(V2 v1, V2 v2)   { return (V2){ v1.x - v2.x, v1.y - v2.y }; }

static inline V2 V2_normal_from_screen(Context *ctx, V2 v) { 
    return V2_divide(v, scale(ctx)); 
}
static inline V2 V2_screen_from_normal(Context *ctx, V2 v) { 
    return V2_scale(v, scale(ctx)); 
}

static void game_compute_screen_collision(Context *ctx, usize i) {
    Body *body = &ctx->bodies.ptr[i];
    f32 norm_width = normal_width(ctx);

    if (body->pos.x - body->radius < 0) {
        body->pos.x = body->radius;
        body->velocity.x *= -game_collision_dampen;
    } else if (body->pos.x + body->radius > norm_width) {
        body->pos.x = norm_width - body->radius;
        body->velocity.x *= -game_collision_dampen;
    }

    if (body->pos.y - body->radius < 0) {
        body->pos.y = body->radius;
        body->velocity.y *= -game_collision_dampen;
    } else if (body->pos.y + body->radius > 1.0f) {
        body->pos.y = 1.0f - body->radius;
        body->velocity.y *= -game_collision_dampen;
    }
}

static void game_render_creator(Context *ctx) {
    Creator creator = ctx->creator;
    Body body = creator.body;

    f32 inner = screen_from_normal(ctx, ctx->cursor_radius);
    f32 outer = inner + screen_from_normal(ctx, creator_ring_thickness);
    if (!creator.active) {
        V2 pos = V2_screen_from_normal(ctx, ctx->mouse_pos);
        DrawRing(pos, inner, outer, 0, 360, 0, colour_creator_inactive);
        return;
    }

    V2 start = V2_screen_from_normal(ctx, body.pos);
    V2 end = V2_screen_from_normal(
        ctx, V2_add(body.pos, creator.displacement)
    );
    f32 line_width = screen_from_normal(ctx, creator_line_width);
    DrawLineEx(start, end, line_width, colour_creator_displacement_line);

    f32 radius = screen_from_normal(ctx, body.radius);
    DrawCircleV(start, radius, colour_creator_active);
}

static void game_update_and_render(Context *ctx) {
    ctx->delta = GetFrameTime();
    ctx->width = GetScreenWidth();
    ctx->height = GetScreenHeight();

    ctx->mouse_pos = V2_normal_from_screen(ctx, GetMousePosition());
    ctx->cursor_radius += GetMouseWheelMove() / 100.0f;
    clamp(ctx->cursor_radius, game_cursor_radius_min, game_cursor_radius_max);

    Creator *creator = &ctx->creator;
    creator->body.mass = powf(ctx->cursor_radius * 1000.0f, 3);
    creator->body.radius = ctx->cursor_radius;

    Array_Body *bodies = &ctx->bodies;
    if (IsKeyPressed('R')) bodies->len = 0;

    if (creator->active) {
        f32 factor = 100.0f;

        if (IsMouseButtonDown(MOUSE_BUTTON_RIGHT)) {
            creator->displacement = V2_sub(ctx->mouse_pos, creator->body.pos);
        } else if (IsMouseButtonReleased(MOUSE_BUTTON_RIGHT)) {
            creator->active = false;
        }

        if (IsMouseButtonPressed(MOUSE_BUTTON_LEFT)) {
            creator->body.velocity = V2_divide(creator->displacement, factor);
            array_push(&ctx->arena, bodies, &creator->body);
        }
    } else {
        if (IsMouseButtonPressed(MOUSE_BUTTON_RIGHT)) {
            creator->active = true;
            creator->displacement = (V2){0};
            creator->body.pos = ctx->mouse_pos;
        } else if (IsMouseButtonPressed(MOUSE_BUTTON_LEFT)) {
            creator->body.pos = ctx->mouse_pos;
            creator->body.velocity = (V2){0};
            array_push(&ctx->arena, bodies, &creator->body);
        }
    }

    for (usize i = 0; i < bodies->len; i += 1) {
        for (usize cmp_i = i + 1; cmp_i < bodies->len; cmp_i += 1) {
            game_compute_interaction(ctx, i, cmp_i);
        }
        game_compute_screen_collision(ctx, i);

        Body *body = &bodies->ptr[i];
        body->pos = V2_add(body->pos, body->velocity);

        V2 pos = V2_screen_from_normal(ctx, body->pos);
        DrawCircleV(pos, screen_from_normal(ctx, body->radius), body->colour);
    }

    game_render_creator(ctx);
}

File: src/Game.zig
const rl = @cImport({
    @cInclude("raylib.h");
    @cInclude("raymath.h");
});
const std = @import("std");

const pow = std.math.pow;
const V2 = @Vector(2, f32);

allocator: std.mem.Allocator,
name: []const u8,
width: c_int,
height: c_int,
fps: c_int = default_fps,
g: f32 = 3e-8 / @as(f32, default_fps),
delta: f32 = 0,
cursor_radius: f32 = 0,
bodies: std.ArrayList(Body) = undefined,
creator: Creator = undefined,
mouse_pos: V2 = .{ 0, 0 },

const default_fps = 60;
const collision_dampen_factor = 0.3;

const Body = struct {
    mass: f32,
    radius: f32,
    pos: V2 = .{ 0, 0 },
    velocity: V2 = .{ 0, 0 },
};

const Creator = struct {
    active: bool = false,
    displacement: V2 = .{ 0, 0 },
    body: Body,

    pub const colour_inactive = Colour.blue;
    pub const colour_active = Colour.grey_dark;
    pub const colour_line = colour_active;
    pub const ring_thickness = 0.004;
    pub const line_width = 0.004;
};

const Colour = struct {
    pub inline fn fromHex(rgb: u24) rl.Color {
        return .{
            .r = (rgb & 0xff0000) >> 16,
            .g = (rgb & 0x00ff00) >> 8,
            .b = (rgb & 0x0000ff),
            .a = 0xff,
        };
    }

    pub const blue = fromHex(0x00008b);
    pub const black = fromHex(0x000000);
    pub const grey_light = fromHex(0xc0c0c0);
    pub const grey_dark = fromHex(0x555555);

    pub const background = grey_light;
    pub const body = black;
};

pub fn init(game: @This()) @This() {
    rl.SetConfigFlags(rl.FLAG_MSAA_4X_HINT | rl.FLAG_WINDOW_RESIZABLE);
    rl.SetTargetFPS(game.fps);
    rl.InitWindow(game.width, game.height, @ptrCast(game.name));

    var result = game;
    result.bodies = std.ArrayList(Body).init(result.allocator);
    return result;
}

pub fn deinit(_: *@This()) void {
    rl.CloseWindow();
}

pub inline fn shouldQuit(_: *@This()) bool {
    return rl.WindowShouldClose();
}

pub fn frameBegin(_: *@This()) void {
    rl.BeginDrawing();
    rl.ClearBackground(Colour.background);
}

pub fn frameEnd(_: *@This()) void {
    rl.EndDrawing();
}

pub fn updateAndRender(self: *@This()) !void {
    self.delta = rl.GetFrameTime();
    self.width = rl.GetScreenWidth();
    self.height = rl.GetScreenHeight();

    self.mouse_pos = self.normalFromScreen(
        v2fromRaylib(rl.GetMousePosition()),
    );
    self.cursor_radius += rl.GetMouseWheelMove() / 100;
    self.cursor_radius = std.math.clamp(self.cursor_radius, 0.01, 0.1);

    const creator = &self.creator;
    creator.body.mass = std.math.pow(f32, self.cursor_radius * 1000, 3);
    creator.body.radius = self.cursor_radius;

    if (rl.IsKeyPressed('R')) self.bodies.shrinkRetainingCapacity(0);

    if (creator.active) {
        if (rl.IsMouseButtonDown(rl.MOUSE_BUTTON_RIGHT)) {
            creator.displacement = self.mouse_pos - creator.body.pos;
        } else if (rl.IsMouseButtonReleased(rl.MOUSE_BUTTON_RIGHT)) {
            creator.active = false;
        }

        if (rl.IsMouseButtonPressed(rl.MOUSE_BUTTON_LEFT)) {
            const factor: V2 = @splat(100);
            creator.body.velocity = creator.displacement / factor;
            try self.bodies.append(creator.body);
        }
    } else {
        if (rl.IsMouseButtonPressed(rl.MOUSE_BUTTON_RIGHT)) {
            creator.active = true;
            creator.displacement = .{ 0, 0 };
            creator.body.pos = self.mouse_pos;
        } else if (rl.IsMouseButtonPressed(rl.MOUSE_BUTTON_LEFT)) {
            creator.body.pos = self.mouse_pos;
            creator.body.velocity = .{ 0, 0 };
            try self.bodies.append(creator.body);
        }
    }

    for (0..self.bodies.items.len) |i| {
        for (i + 1..self.bodies.items.len) |cmp_i| {
            self.computeInteraction(i, cmp_i);
        }
        self.computeScreenCollision(i);

        const body = &self.bodies.items[i];

        body.pos += body.velocity;
        const pos = self.screenFromNormal(body.pos);
        rl.DrawCircleV(
            raylibFromV2(pos),
            self.screenFromNormal(body.radius),
            Colour.black,
        );
    }

    self.renderCreator();
}

fn computeInteraction(self: *@This(), i: usize, cmp_i: usize) void {
    const body = &self.bodies.items[i];
    const body_cmp = &self.bodies.items[cmp_i];

    const dist_xy = body.pos - body_cmp.pos;
    const dist = @sqrt(pow(f32, dist_xy[0], 2) + pow(f32, dist_xy[1], 2));

    const colliding = dist < (body.radius + body_cmp.radius) / 2;
    if (colliding) return;

    const force = -1 * self.delta * self.g *
        body.mass * body_cmp.mass / pow(f32, dist, 2);
    const force_xy = V2{
        force * (dist_xy[0] / dist),
        force * (dist_xy[1] / dist),
    };

    const body_accel = force_xy / @as(V2, @splat(body.mass));
    body.velocity += body_accel;

    const body_cmp_accel = force_xy / @as(V2, @splat(body_cmp.mass));
    body_cmp.velocity -= body_cmp_accel;
}

fn computeScreenCollision(self: *@This(), i: usize) void {
    const body = &self.bodies.items[i];
    const dimensions = [2]f32{ self.normalWidth(), 1 };
    inline for (0..2) |axis| {
        if (body.pos[axis] - body.radius < 0) {
            body.pos[axis] = body.radius;
            body.velocity[axis] *= -collision_dampen_factor;
        } else if (body.pos[axis] + body.radius > dimensions[axis]) {
            body.pos[axis] = dimensions[axis] - body.radius;
            body.velocity[axis] *= -collision_dampen_factor;
        }
    }
}

fn renderCreator(self: @This()) void {
    const creator = self.creator;
    const body = creator.body;

    const inner_radius = self.screenFromNormal(self.cursor_radius);
    const outer_radius = inner_radius +
        self.screenFromNormal(@as(f32, Creator.ring_thickness));
    if (!creator.active) {
        const pos = self.screenFromNormal(self.mouse_pos);
        rl.DrawRing(
            raylibFromV2(pos),
            inner_radius,
            outer_radius,
            0,
            360,
            0,
            Creator.colour_inactive,
        );
        return;
    }

    const line_start = self.screenFromNormal(body.pos);
    const line_end = self.screenFromNormal(body.pos + creator.displacement);
    const line_width = self.screenFromNormal(@as(f32, Creator.line_width));
    rl.DrawLineEx(
        raylibFromV2(line_start),
        raylibFromV2(line_end),
        line_width,
        Creator.colour_line,
    );

    const radius = self.screenFromNormal(body.radius);
    rl.DrawCircleV(
        raylibFromV2(line_start),
        radius,
        Creator.colour_active,
    );
}

inline fn scale(self: @This(), comptime T: type) T {
    const scalar: f32 = @floatFromInt(self.height);
    return switch (T) {
        inline f32 => scalar,
        inline V2 => @as(V2, @splat(scalar)),
        inline else => @compileError("invalid type passed to Game.scale()"),
    };
}

inline fn normalFromScreen(self: @This(), screen: anytype) @TypeOf(screen) {
    return screen / self.scale(@TypeOf(screen));
}

inline fn screenFromNormal(self: @This(), normal: anytype) @TypeOf(normal) {
    return normal * self.scale(@TypeOf(normal));
}

inline fn normalWidth(self: @This()) f32 {
    return @as(f32, @floatFromInt(self.width)) / self.scale(f32);
}

inline fn v2fromRaylib(vector2: rl.Vector2) V2 {
    return .{ vector2.x, vector2.y };
}

inline fn raylibFromV2(v2: V2) rl.Vector2 {
    return .{ .x = v2[0], .y = v2[1] };
}
